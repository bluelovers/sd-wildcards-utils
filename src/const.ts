export const RE_UNSAFE_QUOTE = /['"]/;
export const RE_UNSAFE_QUOTE_DOUBLE = /"/;
export const RE_UNSAFE_VALUE = /^\s*-|[{$~!@}\n|:?#'"%]/;
//export const RE_UNSAFE_PLAIN = /-/;
export const RE_UNSAFE_PLAIN = /-/;
