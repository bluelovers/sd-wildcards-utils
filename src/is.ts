import { isDocument, YAMLMap, Document, isMap } from 'yaml';
import { IWildcardsYAMLDocument, IWildcardsYAMLMapRoot, IWildcardsYAMLPairValue, IWildcardsYAMLScalar } from './types';

export function isWildcardsYAMLDocument<T extends IWildcardsYAMLMapRoot>(node: IWildcardsYAMLDocument<T, true> | Document<T, true>): node is IWildcardsYAMLDocument<T, true>
export function isWildcardsYAMLDocument<T extends IWildcardsYAMLDocument | Document>(doc: any): doc is IWildcardsYAMLDocument
export function isWildcardsYAMLDocument<T extends YAMLMap = IWildcardsYAMLMapRoot>(node: any): node is IWildcardsYAMLDocument<T, true>
export function isWildcardsYAMLDocument(doc: any)
{
	return isDocument(doc)
}

export function isWildcardsYAMLDocumentAndContentsIsMap(doc: any): doc is IWildcardsYAMLDocument
{
	return isDocument(doc) && isMap(doc.contents)
}

export function isWildcardsYAMLMap<K extends IWildcardsYAMLScalar, V extends IWildcardsYAMLPairValue>(doc: IWildcardsYAMLMapRoot<K, V> | YAMLMap.Parsed<K, V> | YAMLMap<K, V>): doc is IWildcardsYAMLMapRoot<K, V>
export function isWildcardsYAMLMap<K extends IWildcardsYAMLScalar = IWildcardsYAMLScalar, V extends IWildcardsYAMLPairValue = IWildcardsYAMLPairValue>(doc: any): doc is IWildcardsYAMLMapRoot<K, V>
export function isWildcardsYAMLMap(doc: any)
{
	return isMap(doc)
}
