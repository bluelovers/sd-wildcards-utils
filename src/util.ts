import { array_unique_overwrite } from 'array-hyper-unique';
import {
	IMatchDynamicPromptsWildcardsEntry,
	IVisitPathsNode,
	IYamlNodeTypeSymbol
} from './types';

export const SYMBOL_YAML_NODE_TYPE_ALIAS = Symbol.for('yaml.alias');
export const SYMBOL_YAML_NODE_TYPE_DOC = Symbol.for('yaml.document');
export const SYMBOL_YAML_NODE_TYPE_MAP = Symbol.for('yaml.map');
export const SYMBOL_YAML_NODE_TYPE_PAIR = Symbol.for('yaml.pair');
export const SYMBOL_YAML_NODE_TYPE_SCALAR = Symbol.for('yaml.scalar');
export const SYMBOL_YAML_NODE_TYPE_SEQ = Symbol.for('yaml.seq');
export const SYMBOL_YAML_NODE_TYPE = Symbol.for('yaml.node.type');

export const RE_DYNAMIC_PROMPTS_WILDCARDS = /(?<!#[^\n]*)__([&~!@])?([*\w\/_\-]+)(\([^\n#]+\))?__/

/**
 * for `matchAll`
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/matchAll
 */
export const RE_DYNAMIC_PROMPTS_WILDCARDS_GLOBAL = new RegExp(RE_DYNAMIC_PROMPTS_WILDCARDS, RE_DYNAMIC_PROMPTS_WILDCARDS.flags + 'g')

export const RE_WILDCARDS_NAME = /^[\w\-_\/]+$/

/**
 * Checks if the input string matches the dynamic prompts wildcards pattern.
 *
 * @param input - The input string to check.
 * @returns A boolean indicating whether the input string matches the pattern.
 *
 * @remarks
 * This function uses the `matchDynamicPromptsWildcards` function to perform the check.
 * It returns `true` if the input string is a full match, and `false` otherwise.
 *
 * @example
 * ```typescript
 * const input1 = "__season_clothes(season=winter)__";
 * console.log(isDynamicPromptsWildcards(input1)); // Output: true
 *
 * const input2 = "__season_clothes(season=__season_clothes__)__";
 * console.log(isDynamicPromptsWildcards(input2)); // Output: true
 *
 * const input3 = "This is not a wildcards pattern";
 * console.log(isDynamicPromptsWildcards(input3)); // Output: false
 * ```
 */
export function isDynamicPromptsWildcards(input: string): boolean
{
	return matchDynamicPromptsWildcards(input).isFullMatch;
}

/**
 * Matches the input string against the dynamic prompts wildcards pattern.
 *
 * @see https://github.com/adieyal/sd-dynamic-prompts/blob/main/docs/SYNTAX.md
 *
 * @param input - The input string to match.
 * @returns An object containing the matched groups or `null` if no match is found.
 *
 * @remarks
 * This function uses the `RE_DYNAMIC_PROMPTS_WILDCARDS` regular expression to perform the match.
 * The returned object contains the following properties:
 * - `name`: The name extracted from the input string.
 * - `variables`: The variables extracted from the input string.
 * - `keyword`: The keyword extracted from the input string.
 * - `source`: The original matched source string.
 * - `isFullMatch`: A boolean indicating whether the input string is a full match.
 *
 * @example
 * ```typescript
 * const input = "\_\_season_clothes(season=winter)\_\_";
 * const result = matchDynamicPromptsWildcards(input);
 * console.log(result);
 * // Output: { name: 'season_clothes', variables: '(season=winter)', keyword: undefined, source: '\__season_clothes(season=winter)\__', isFullMatch: true }
 * ```
 *
 * @example
 * __season_clothes(season=winter)__
 * __season_clothes(season=__season_clothes__)__
 * __season_clothes(season=!__season_clothes__)__
 *
 * __season_clothes(season=__@season_clothes__)__
 * __season_clothes(season=__~season_clothes__)__
 *
 * __@season_clothes(season=__season_clothes__)__
 * __~season_clothes(season=__season_clothes__)__
 *
 * __season_clothes(season={summer|autumn|winter|spring})__
 * __season_clothes(season=!{summer|autumn|winter|spring})__
 *
 * __season_clothes(season={@summer|autumn|winter|spring})__
 * __season_clothes(season={!summer|autumn|winter|spring})__
 *
 * __season_clothes(season=)__
 */
export function matchDynamicPromptsWildcards(input: string)
{
	const m = input.match(RE_DYNAMIC_PROMPTS_WILDCARDS);
	return _matchDynamicPromptsWildcardsCore(m, input);
}

export function _matchDynamicPromptsWildcardsCore(m: RegExpMatchArray,
	input?: string,
): IMatchDynamicPromptsWildcardsEntry
{
	if (!m) return null;

	let [source, keyword, name, variables] = m;

	return {
		name,
		variables,
		keyword,
		source,
		isFullMatch: source === (input ?? m.input),
		isStarWildcards: name.includes('*'),
	}
}

/**
 * Generator function that matches all occurrences of the dynamic prompts wildcards pattern in the input string.
 */
export function* matchDynamicPromptsWildcardsAllGenerator(input: string)
{
	const ls = input.matchAll(RE_DYNAMIC_PROMPTS_WILDCARDS_GLOBAL);

	for (let m of ls)
	{
		yield _matchDynamicPromptsWildcardsCore(m, input);
	}
}

/**
 * Converts the generator function `matchDynamicPromptsWildcardsAllGenerator` into an array.
 */
export function matchDynamicPromptsWildcardsAll(input: string, unique?: boolean)
{
	const arr = [...matchDynamicPromptsWildcardsAllGenerator(input)] as IMatchDynamicPromptsWildcardsEntry[];

	return unique ? array_unique_overwrite(arr) : arr
}

/**
 * Checks if the given name is a valid Wildcards name.
 *
 * @param name - The name to check.
 * @returns A boolean indicating whether the name is valid.
 *
 * @remarks
 * A valid Wildcards name should:
 * - Only contain alphanumeric characters, hyphens, or underscores.
 * - Not start or end with an underscore.
 * - Not contain consecutive underscores.
 *
 * @example
 * ```typescript
 * const name1 = "season_clothes";
 * console.log(isWildcardsName(name1)); // Output: true
 *
 * const name2 = "_season_clothes";
 * console.log(isWildcardsName(name2)); // Output: false
 *
 * const name3 = "season_clothes_";
 * console.log(isWildcardsName(name3)); // Output: false
 *
 * const name4 = "season__clothes";
 * console.log(isWildcardsName(name4)); // Output: false
 *
 * const name5 = "season-clothes";
 * console.log(isWildcardsName(name5)); // Output: true
 * ```
 */
export function isWildcardsName(name: string): boolean
{
	return RE_WILDCARDS_NAME.test(name) && !/__|[_\/]$|^[_\/]|\/\//.test(name)
}

export function assertWildcardsName(name: string)
{
	if (isWildcardsName(name))
	{
		throw new SyntaxError(`Invalid Wildcards Name Syntax: ${name}`)
	}
}

export function convertWildcardsNameToPaths(name: string)
{
	return name.split('/');
}

export function isWildcardsPathSyntx(path: string): path is `__${string}__`
{
	return RE_DYNAMIC_PROMPTS_WILDCARDS.test(path)
}

export function wildcardsPathToPaths(path: string)
{
	if (isWildcardsPathSyntx(path))
	{
		path = matchDynamicPromptsWildcards(path).name
	}

	return convertWildcardsNameToPaths(path);
}

export function getNodeTypeSymbol(node: IVisitPathsNode): IYamlNodeTypeSymbol
{
	// @ts-ignore
	return node?.[SYMBOL_YAML_NODE_TYPE]
}

export function _getNodeTypeCore(sym: IYamlNodeTypeSymbol)
{
	try
	{
		return Symbol.keyFor(sym)
	}
	catch (e)
	{

	}
}

export function getNodeType(node: IVisitPathsNode)
{
	return _getNodeTypeCore(getNodeTypeSymbol(node))
}

export function isSameNodeType(a: IVisitPathsNode, b: IVisitPathsNode)
{
	const s = getNodeTypeSymbol(a);

	return s && getNodeTypeSymbol(b) === s;
}
