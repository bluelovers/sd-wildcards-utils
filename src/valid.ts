import { Document, isDocument, isMap, isNode, YAMLMap } from 'yaml';
import { IWildcardsYAMLDocument, visitWildcardsYAML } from './items';
import { IOptionsSharedWildcardsYaml } from './options';
import { IRecordWildcards } from './index';

export function _validMap(key: number | 'key' | 'value' | null, node: YAMLMap)
{
	const elem = node.items.find(pair => pair.value === null);
	if (elem)
	{
		throw new SyntaxError(`Invalid SYNTAX. key: ${key}, node: ${node}`)
	}
}

export function validWildcardsYamlData<T extends IRecordWildcards | IWildcardsYAMLDocument | Document>(data: T | unknown,
	opts?: IOptionsSharedWildcardsYaml,
): asserts data is T
{
	if (isDocument(data))
	{
		if (isNode(data.contents) && !isMap(data.contents))
		{
			throw TypeError(`The 'contents' property of the provided YAML document must be a YAMLMap. Received: ${data.contents}`)
		}

		visitWildcardsYAML(data, {
			Map: _validMap,
		});

		data = data.toJSON()
	}

	let rootKeys = Object.keys(data);

	if (!rootKeys.length)
	{
		throw TypeError(`The provided JSON contents must contain at least one key.`)
	}
	else if (rootKeys.length !== 1 && !opts?.allowMultiRoot)
	{
		throw TypeError(`The provided JSON object cannot have more than one root key. Only one root key is allowed unless explicitly allowed by the 'allowMultiRoot' option.`)
	}
}
