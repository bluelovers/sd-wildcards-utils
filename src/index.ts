/**
 * Created by user on 2024/5/21.
 */
import { Document, isDocument, ParsedNode, parseDocument, stringify, YAMLMap } from 'yaml';
import {
	defaultOptionsParseDocument,
	defaultOptionsStringify,
	getOptionsFromDocument,
} from './options';
import { _visitNormalizeScalar, visitWildcardsYAML } from './node/node-items';
import { createDefaultVisitWildcardsYAMLOptions, validWildcardsYamlData } from './valid';
import {
	IOptionsParseDocument,
	IOptionsStringify,
	IOptionsVisitor,
	IParseWildcardsYamlInputSource,
	IRecordWildcards,
	IWildcardsYAMLDocument,
	IWildcardsYAMLDocumentParsed,
	IWildcardsYAMLMapRoot,
} from './types';
import { _expandForwardSlashKeys } from './parser';

export * from './util';
export * from './options';
export * from './node/node-items';
export * from './valid';
export * from './prompts/valid-prompts';
export * from './node/node-merge';
export * from './node/node-find';
export * from './prompts/format';
export * from './prompts/prompts';
export * from './check';
export * from './node/node-is';
export type * from './types';

/**
 * Normalizes a YAML document by applying specific rules to its nodes.
 **/
export function normalizeDocument<T extends Document>(doc: T, opts?: IOptionsParseDocument)
{
	let options = getOptionsFromDocument(doc, opts);

	const defaults = createDefaultVisitWildcardsYAMLOptions(options);

	let checkUnsafeQuote = !options.disableUnsafeQuote;

	let visitorOptions: IOptionsVisitor = {
		...defaults,

		Scalar(key, node, parentNodes)
		{
			return _visitNormalizeScalar(key, node, parentNodes, {
				checkUnsafeQuote,
				options,
			})
		},
	};

	visitWildcardsYAML(doc, visitorOptions)
}

/**
 * Converts the given YAML data to a string, applying normalization and formatting.
 *
 * @returns - A string representation of the input YAML data, with normalization and formatting applied.
 *
 * @throws - Throws a `SyntaxError` if the input data is invalid according to the `validWildcardsYamlData` function.
 *
 * @remarks
 * This function takes the input YAML data and applies normalization and formatting using the provided options.
 * If the input data is a `Document` object, it first normalizes the document using the `normalizeDocument` function.
 * Then, it converts the normalized document to a string using the `toString` method with the specified options.
 * If the input data is not a `Document` object, it directly converts the data to a string using the `stringify` function with the specified options.
 *
 * @example
 * ```typescript
 * const yamlData: IRecordWildcards = {
 *   key1: ['value1', 'value2'],
 *   key2: {
 *     subkey1: ['value3', 'value4'],
 *   },
 * };
 *
 * const yamlString = stringifyWildcardsYamlData(yamlData);
 * console.log(yamlString);
 * // Output:
 * // key1:
 * //   - value1
 * //   - value2
 * // key2:
 * //   subkey1:
 * //     - value3
 * //     - value4
 * ```
 */
export function stringifyWildcardsYamlData<T extends IRecordWildcards | IWildcardsYAMLDocument | Document>(data: T | unknown,
	opts?: IOptionsStringify,
)
{
	const isDoc = isDocument(data);

	if (isDoc)
	{
		opts = getOptionsFromDocument(data, opts);
	}

	opts = defaultOptionsStringify(opts);

	if (isDoc)
	{
		normalizeDocument(data, opts);

		return data.toString(opts)
	}

	return stringify(data, opts)
}

/**
 * Parses Stable Diffusion wildcards source to a YAML object.
 *
 * @returns - If `Contents` extends `ParsedNode`, returns a parsed `Document.Parsed` with the specified `Contents` and `Strict`.
 *            Otherwise, returns a parsed `Document` with the specified `Contents` and `Strict`.
 *
 * @throws - Throws a `SyntaxError` if the YAML data is invalid according to the `validWildcardsYamlData` function.
 *
 * @remarks
 * This function parses the given `source` string or Uint8Array to a YAML object.
 * It uses the `parseDocument` function from the `yaml` library with `keepSourceTokens: true` option.
 * Then, it validates the parsed data using the `validWildcardsYamlData` function.
 * Finally, it returns the parsed data.
 */
export function parseWildcardsYaml<Contents extends YAMLMap = IWildcardsYAMLMapRoot, Strict extends boolean = true>(source: IParseWildcardsYamlInputSource,
	opts?: IOptionsParseDocument,
): Contents extends ParsedNode
	? IWildcardsYAMLDocumentParsed<Contents, Strict>
	: IWildcardsYAMLDocument<Contents, Strict>
{
	opts = defaultOptionsParseDocument(opts);

	if (opts.allowEmptyDocument)
	{
		source ??= '';
	}

	let data = parseDocument<Contents, Strict>(source.toString(), opts);

	// expand keys with forward slashes into nested maps
	if (opts.expandForwardSlashKeys)
	{
		_expandForwardSlashKeys(data as Document);
	}

	validWildcardsYamlData(data, opts);

	return data as any
}

export default parseWildcardsYaml

