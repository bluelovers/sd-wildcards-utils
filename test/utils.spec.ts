//@noUnusedParameters:false
/// <reference types="jest" />
/// <reference types="node" />
/// <reference types="expect" />

import { isWildcardsName, matchDynamicPromptsWildcards } from '../src/util';
import parseWildcardsYaml, { getOptionsFromDocument } from '../src/index';
import { _checkValue } from '../src/valid';

beforeAll(async () =>
{

});

test(`doc.options`, () => {
	let doc = parseWildcardsYaml(null, {
		allowEmptyDocument: true,
	});

	expect(getOptionsFromDocument(doc)).toMatchSnapshot()
});

describe(`matchDynamicPromptsWildcards`, () =>
{

	test.skip(`dummy`, () => {});

	test.each([
		'__season_clothes(season=winter)__',
		'__season_clothes__',
		'__season_clothes(season=__season_clothes__)__',

		'__season_clothes(season=!__season_clothes__)__',

		'__~season_clothes(season=winter)__',

		'__season_clothes(season=__@season_clothes__)__',
		'__season_clothes(season=__~season_clothes__)__',

		'__@season_clothes(season=__season_clothes__)__',
		'__~season_clothes(season=__season_clothes__)__',

		'__season_clothes(season={summer|autumn|winter|spring})__',
		'__season_clothes(season=!{summer|autumn|winter|spring})__',

		'__season_clothes(season={@summer|autumn|winter|spring})__',
		'__season_clothes(season={!summer|autumn|winter|spring})__',

		'__season_clothes(season=!{@summer|autumn|winter|spring})__',
		'__season_clothes(season=!{!summer|autumn|winter|spring})__',

		'__season_clothes(season=)__',

		'__season_clothes(season=)__ ',

		' __season_clothes(season=)__ ',

		' __season_clothes(season= )__ ',

		// ---------------

		'__scope/season_clothes(season=winter)__',
		'__scope/season_clothes__',
		'__scope/season_clothes(season=__scope/season_clothes__)__',

		'__scope/season_clothes(season=!__scope/season_clothes__)__',

		'__~scope/season_clothes(season=winter)__',

		'__scope/season_clothes(season=__@scope/season_clothes__)__',
		'__scope/season_clothes(season=__~scope/season_clothes__)__',

		'__@scope/season_clothes(season=__scope/season_clothes__)__',
		'__~scope/season_clothes(season=__scope/season_clothes__)__',

		'__scope/season_clothes(season={summer|autumn|winter|spring})__',
		'__scope/season_clothes(season=!{summer|autumn|winter|spring})__',

		'__scope/season_clothes(season={@summer|autumn|winter|spring})__',
		'__scope/season_clothes(season={!summer|autumn|winter|spring})__',

		'__scope/season_clothes(season=!{@summer|autumn|winter|spring})__',
		'__scope/season_clothes(season=!{!summer|autumn|winter|spring})__',

		'__scope/season_clothes(season=)__',

		'__scope/season_clothes(season=)__ ',

		' __scope/season_clothes(season=)__ ',

		' __scope/season_clothes(season= )__ ',

	])(`%j`, (input) =>
	{

		let actual = matchDynamicPromptsWildcards(input);

		expect(actual).toMatchSnapshot();

		expect(isWildcardsName(actual.name)).toBeTruthy();

	});

})

describe(`_checkValue`, () => {

	describe(`valid`, () => {

		test.each([
			'2b_\(nier:automata\)_\(cosplay\)',
			'purple_gray',
		])(`%j`, (input) => {

			let actual = _checkValue(input);

			expect(actual).toBeUndefined();
		})

	})

	describe(`invalid`, () => {

		test.each([
			' __lazy-wildcards/subject/env-elem/stairs/prompts_ ',
			' __lazy-wildcards/subject/__env-elem/stairs/prompts__ ',
			' {__lazy-wildcards/subject/env-elem/stairs/prompts_} ',
			' _lazy-wildcards/subject/env-elem/stairs/prompts__ ',
			' __lazy-wildcards/subject/env-elem__/stairs/prompts__ ',
			' {_lazy-wildcards/subject/env-elem/stairs/prompts__} ',
		])(`%j`, (input) => {

			let actual = _checkValue(input);

			expect(actual).not.toBeUndefined();
			expect(actual).toMatchSnapshot();
		})

	})

})
