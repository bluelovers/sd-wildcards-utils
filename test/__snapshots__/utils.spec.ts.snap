// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doc.options 1`] = `
{
  "_directives": Directives {
    "atNextDocument": false,
    "docEnd": false,
    "docStart": null,
    "tags": {
      "!!": "tag:yaml.org,2002:",
    },
    "yaml": {
      "explicit": false,
      "version": "1.2",
    },
  },
  "allowEmptyDocument": true,
  "intAsBigInt": false,
  "keepSourceTokens": true,
  "logLevel": "warn",
  "prettyErrors": true,
  "strict": true,
  "toStringDefaults": {
    "allowMultiRoot": undefined,
    "blockQuote": true,
    "collectionStyle": "block",
    "defaultKeyType": "PLAIN",
    "defaultStringType": "PLAIN",
    "disableUniqueItemValues": undefined,
    "disableUnsafeQuote": undefined,
    "minifyPrompts": undefined,
  },
  "uniqueKeys": true,
  "version": "1.2",
}
`;

exports[`matchDynamicPromptsWildcards " __scope/season_clothes(season= )__ " 1`] = `
{
  "isFullMatch": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season= )__",
  "variables": "(season= )",
}
`;

exports[`matchDynamicPromptsWildcards " __scope/season_clothes(season=)__ " 1`] = `
{
  "isFullMatch": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=)__",
  "variables": "(season=)",
}
`;

exports[`matchDynamicPromptsWildcards " __season_clothes(season= )__ " 1`] = `
{
  "isFullMatch": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season= )__",
  "variables": "(season= )",
}
`;

exports[`matchDynamicPromptsWildcards " __season_clothes(season=)__ " 1`] = `
{
  "isFullMatch": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=)__",
  "variables": "(season=)",
}
`;

exports[`matchDynamicPromptsWildcards "__@scope/season_clothes(season=__scope/season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": "@",
  "name": "scope/season_clothes",
  "source": "__@scope/season_clothes(season=__scope/season_clothes__)__",
  "variables": "(season=__scope/season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__@season_clothes(season=__season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": "@",
  "name": "season_clothes",
  "source": "__@season_clothes(season=__season_clothes__)__",
  "variables": "(season=__season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__~scope/season_clothes(season=__scope/season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": "~",
  "name": "scope/season_clothes",
  "source": "__~scope/season_clothes(season=__scope/season_clothes__)__",
  "variables": "(season=__scope/season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__~scope/season_clothes(season=winter)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": "~",
  "name": "scope/season_clothes",
  "source": "__~scope/season_clothes(season=winter)__",
  "variables": "(season=winter)",
}
`;

exports[`matchDynamicPromptsWildcards "__~season_clothes(season=__season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": "~",
  "name": "season_clothes",
  "source": "__~season_clothes(season=__season_clothes__)__",
  "variables": "(season=__season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__~season_clothes(season=winter)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": "~",
  "name": "season_clothes",
  "source": "__~season_clothes(season=winter)__",
  "variables": "(season=winter)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=!__scope/season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=!__scope/season_clothes__)__",
  "variables": "(season=!__scope/season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=!{!summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=!{!summer|autumn|winter|spring})__",
  "variables": "(season=!{!summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=!{@summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=!{@summer|autumn|winter|spring})__",
  "variables": "(season=!{@summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=!{summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=!{summer|autumn|winter|spring})__",
  "variables": "(season=!{summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=)__ " 1`] = `
{
  "isFullMatch": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=)__",
  "variables": "(season=)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=)__",
  "variables": "(season=)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=__@scope/season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=__@scope/season_clothes__)__",
  "variables": "(season=__@scope/season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=__~scope/season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=__~scope/season_clothes__)__",
  "variables": "(season=__~scope/season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=__scope/season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=__scope/season_clothes__)__",
  "variables": "(season=__scope/season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season={!summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season={!summer|autumn|winter|spring})__",
  "variables": "(season={!summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season={@summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season={@summer|autumn|winter|spring})__",
  "variables": "(season={@summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season={summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season={summer|autumn|winter|spring})__",
  "variables": "(season={summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=winter)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=winter)__",
  "variables": "(season=winter)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes__",
  "variables": undefined,
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=!__season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=!__season_clothes__)__",
  "variables": "(season=!__season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=!{!summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=!{!summer|autumn|winter|spring})__",
  "variables": "(season=!{!summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=!{@summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=!{@summer|autumn|winter|spring})__",
  "variables": "(season=!{@summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=!{summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=!{summer|autumn|winter|spring})__",
  "variables": "(season=!{summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=)__ " 1`] = `
{
  "isFullMatch": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=)__",
  "variables": "(season=)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=)__",
  "variables": "(season=)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=__@season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=__@season_clothes__)__",
  "variables": "(season=__@season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=__~season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=__~season_clothes__)__",
  "variables": "(season=__~season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=__season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=__season_clothes__)__",
  "variables": "(season=__season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season={!summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season={!summer|autumn|winter|spring})__",
  "variables": "(season={!summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season={@summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season={@summer|autumn|winter|spring})__",
  "variables": "(season={@summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season={summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season={summer|autumn|winter|spring})__",
  "variables": "(season={summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=winter)__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=winter)__",
  "variables": "(season=winter)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes__" 1`] = `
{
  "isFullMatch": true,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes__",
  "variables": undefined,
}
`;
